pkg/api/core/v1/client/portforwarder.go

```diff
@@ -0,0 +1,304 @@
+				ip := ListenAddress{Address: "::1", Protocol: "tcp6", FailureMode: "all"}
+				parsed[ip.Address] = ip
+			}
+		} else if net.ParseIP(address).To4() != nil {
+			parsed[address] = ListenAddress{Address: address, Protocol: "tcp4", FailureMode: "any"}
+		} else if net.ParseIP(address) != nil {
+			parsed[address] = ListenAddress{Address: address, Protocol: "tcp6", FailureMode: "any"}
+		} else {
+			return nil, fmt.Errorf("%s is not a valid IP", address)
+		}
+	}
+	addresses = make([]ListenAddress, len(parsed))
+	id := 0
+	for _, v := range parsed {
+		addresses[id] = v
+		id++
+	}
+	// Sort addresses before returning to get a stable order
+	sort.Slice(addresses, func(i, j int) bool { return addresses[i].Address < addresses[j].Address })
+
+	return addresses, nil
+}
+
+/*
+valid port specifications:
+
+5000
+- forwards from localhost:5000 to pod:5000
+
+8888:5000
+- forwards from localhost:8888 to pod:5000
+
+0:5000
+:5000
+  - selects a random available local port,
+    forwards from localhost:<random port> to pod:5000
+*/
+func parsePorts(ports []string) ([]ForwardedPort, error) {
+	var forwards []ForwardedPort
+	for _, portString := range ports {
+		parts := strings.Split(portString, ":")
+		var localString, remoteString string
+		if len(parts) == 1 {
+			localString = parts[0]
+			remoteString = parts[0]
+		} else if len(parts) == 2 {
+			localString = parts[0]
+			if localString == "" {
+				// support :5000
+				localString = "0"
+			}
+			remoteString = parts[1]
+		} else {
+			return nil, fmt.Errorf("invalid port format '%s'", portString)
+		}
+
+		localPort, err := strconv.ParseUint(localString, 10, 16)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing local port '%s': %s", localString, err)
+		}
+
+		remotePort, err := strconv.ParseUint(remoteString, 10, 16)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing remote port '%s': %s", remoteString, err)
+		}
+		if remotePort == 0 {
+			return nil, fmt.Errorf("remote port must be > 0")
+		}
+
+		forwards = append(forwards, ForwardedPort{uint16(localPort), uint16(remotePort)})
+	}
+
+	return forwards, nil
+}
+
+func (pf *ServicePortForwarder) ForwardPorts() error {
+	return pf.forward()
+}
+
+func (pf *ServicePortForwarder) forward() error {
+	var err error
+
+	listenSuccess := false
+	for i := range pf.Ports {
+		port := &pf.Ports[i]
+		err = pf.listenOnPort(port)
+		switch {
+		case err == nil:
+			listenSuccess = true
+		default:
+			pf.Client.log.V(1).Info("Unable to listen on port %d: %v\n", port.Local)
+			return err
+		}
+	}
+
+	if !listenSuccess {
+		return fmt.Errorf("unable to listen on any of the requested ports: %v", pf.Ports)
+	}
+
+	// wait for interrupt or conn closure
+	<-pf.StopChan
+	pf.Client.log.V(1).Info("lost connection to pod")
+
+	return nil
+}
+
+func (pf *ServicePortForwarder) listenOnPort(port *ForwardedPort) error {
+	var errors []error
+	failCounters := make(map[string]int, 2)
+	successCounters := make(map[string]int, 2)
```

Comment:
I'm a bit confused about the logic and the usage of this counters.
If I understood correctly we are spawing goroutines for every addresses, and start listening.
Maybe we can achieve the same using waitgroups and checking the returned errors.

Reply:
These are simply maps used to count how many failures/successes there are after the user has started epinio client.
They are used to check if the initial listening asset has properly initialized.
This is supposed to fail sync when the user starts epinio client.
I don't see many much added value changing this from this implementation.